src/routes/auth/index.ts

```diff
@@ -0,0 +1,147 @@
+import sgMail from '@sendgrid/mail';
+import { type Context, Hono } from 'hono';
+import { ZodError } from 'zod';
+import { generateEmailHtml } from '../../email-templates/confirm-email.ts';
+import { hashPasswordPBKDF2 } from '../../utils/hashPassword.ts';
+import { StatusCodes, type StatusResponse } from '../../utils/responses.ts';
+import { insertProfile } from '../profile/data.ts';
+import { type CreateProfileRequestBody, CreateProfileSchema } from '../profile/validation.ts';
+import { activateProfile, checkEmail, getPassword } from './data.ts';
+import { SignInSchema } from './validate.ts';
+
+export const authRoutes = new Hono();
+
+authRoutes.post('/sign-up', async (context: Context<EnvironmentBindings>) => {
+	const body = await context.req.json();
+	let parsedBody: CreateProfileRequestBody;
+
+	try {
+		parsedBody = CreateProfileSchema.parse(body);
+	} catch (error) {
+		return context.json<StatusResponse>({ message: error?.message ?? 'Invalid input' }, StatusCodes.BAD_REQUEST);
+	}
+
+	const emailExists = await checkEmail(context.env.database, parsedBody.email);
+	if (emailExists) {
+		return context.json({ message: 'Duplicate email' }, StatusCodes.CONFLICT);
+	}
+
+	// Create Profile
+	try {
+		// Generate salt (16 bytes)
+		const salt = crypto.getRandomValues(new Uint8Array(16));
+
+		// Hash password with PBKDF2
+		const hashedPassword = await hashPasswordPBKDF2(parsedBody.password, salt);
+
+		// Convert salt to Base64 for storage
+		const saltBase64 = btoa(String.fromCharCode(...salt));
+
+		// Store the hashed password and salt in the database
+		parsedBody.password = `${saltBase64}:${hashedPassword}`;
+		await insertProfile({ payload: parsedBody, database: context.env.database });
+	} catch (error) {
+		console.error('Error creating profile', error);
+		return context.json<StatusResponse>({ message: 'Error creating profile' }, StatusCodes.INTERNAL_SERVER_ERROR);
+	}
```

Comment:
Curious why you're catching unspecific errors. Wouldn't the default behaviour in Hono to return a `500` ? Don't catch unless you have a reason to.

Reply:
I think this is my fault, I did follow this pattern before.

As for default error catching behavior in Hono: I'm not sure how it does handle errors and exceptions, the docs are lacking in this regard...

Reply:
I think you should generally not try to catch every error, but instead build a middleware higher up the chain that catches these and returns `500` for anything unexpected + anything else you want to do with the error.

Having all these different spots that swallow the underlying problem and send a blank 500 just causes you to lose information about bugs. So I would remove all unspecific catch blocks, which as a plus makes a lot of the controller code simpler and reduces a nesting level.

Reply:
Yes, that does make sense to have a "catch all" middleware at the very end.

I was thinking of having it more localized per controller, but yeah. It is better the way you are suggesting.

On a side note: I tested hono and it just sends a simple 500 error with a plain text body of "internal server error" for uncaught exceptions.
This is good enough 🤷‍♂️